### DISCUSSION TITLE
Store Paths In Dataset After Moving.

### DISCUSSION THREAD

**Discussion by xiangzhi** [microsoft/psi#192] [Q&A] [answered]
Hello,

I have been using datasets as a way to store data in my experiment (Each participant has their own dataset with each trial being its separate session). One of the issue I have ran into is that I cannot easily copy/move the dataset to another location of my drive. 

Currently, all store paths are save with their static path and moving/copying it will break all the links. Is there a way around it besides creating a new dateset file at the new location?

**Comment by sandrist** [comment, marked as answer]
Hey Zhi,

The store paths are saved as relative paths to the location of the dataset file, so if you're moving the entire dataset together (including the file and all stores), it should just work. If you instead were hoping to move *just* the dataset file, then yes, you'll need to reconstruct a new dataset file since the relative paths will have changed (but hopefully that's not too hard to do).

**Reply by xiangzhi** [reply]
Hey Sean,

Thanks for the quick reply! I just tested it again and the store paths were saved as relative paths when it was first created.
```
{
  "Name": "x1",
  "Sessions": [
    {
      "Name": "x1.first_store.0",
      "Partitions": [
        {
          "$type": "Microsoft.Psi.Data.Partition`1[[Microsoft.Psi.Data.IStreamReader, Microsoft.Psi]], Microsoft.Psi.Data",
          "Name": "recording",
          "StoreName": "recording",
          "StorePath": "recording\\first_store\\recording.0000",
          "StreamReaderTypeName": "Microsoft.Psi.Data.PsiStoreStreamReader, Microsoft.Psi, Version=0.15.49.1, Culture=neutral, PublicKeyToken=null"
        }
      ]
    }
  ]
}
```
However, the moment I add a new session to the dataset, the paths became static.
```
{
  "Name": "x1",
  "Sessions": [
    {
      "Name": "x1.first_store.0",
      "Partitions": [
        {
          "$type": "Microsoft.Psi.Data.Partition`1[[Microsoft.Psi.Data.IStreamReader, Microsoft.Psi]], Microsoft.Psi.Data",
          "Name": "recording",
          "StoreName": "recording",
          "StorePath": "E:\\Study-Data\\x1\\recording\\first_store\\recording.0000",
          "StreamReaderTypeName": "Microsoft.Psi.Data.PsiStoreStreamReader, Microsoft.Psi, Version=0.15.49.1, Culture=neutral, PublicKeyToken=null"
        }
      ]
    },
    {
      "Name": "x1.second_store.0",
      "Partitions": [
        {
          "$type": "Microsoft.Psi.Data.Partition`1[[Microsoft.Psi.Data.IStreamReader, Microsoft.Psi]], Microsoft.Psi.Data",
          "Name": "recording",
          "StoreName": "recording",
          "StorePath": "E:\\Study-Data\\x1\\recording\\second_store\\recording.0000",
          "StreamReaderTypeName": "Microsoft.Psi.Data.PsiStoreStreamReader, Microsoft.Psi, Version=0.15.49.1, Culture=neutral, PublicKeyToken=null"
        }
      ]
    }
  ]
}
```
This might be a bug? The only difference is that creation vs loading of a Dataset.

**Reply by xiangzhi** [reply]
I think I found the culprit.

When I load an existing datasets with:
```
dataset = Dataset.Load(datasetPath, autoSave: true);
```
The dataset is loaded with `useRelativePaths` set to `false`.  I'm not exactly sure why it is set to `false`. The dataset file does not have a value for `useRelativePaths` and the JSON constructor just passes `false`, ignoring the code's default. Subsequently, the dataset is saved with `useRelativePaths` set to `false`.

You should be able to reproduce this by adding the following line to the `DatasetRelativePaths` function in `DatasetTests.cs`:
```

            // reload the saved dataset and verify that store paths are still valid
            var newDataset = Dataset.Load(newDatasetFile);
            Assert.IsTrue(newDataset.Sessions[0].Partitions[0].AvailableStreams.Count() > 0);

            // Check if relative path is set to true for the new dataset.
            Assert.IsTrue(newDataset.UseRelativePaths);
        }
```

**Reply by sandrist** [reply]
Interesting, that might indeed be a bug. We'll take a look.

**Reply by xiangzhi** [reply]
Awesome, thanks!

**Reply by skdent** [reply]
Hi Xi

I've updated the Dataset code so that Stores within it are always specified with relative paths if they reside on the same drive as the Dataset.

https://github.com/microsoft/psi/pull/201

**Reply by xiangzhi** [reply]
Yeah! Thanks for the fix!!
